OpenRA.Mods.Common/SpriteLoaders/GifLoader.cs

```diff
@@ -0,0 +1,315 @@
+
+								var identifier = s.ReadASCII(8);
+								s.Position += 3;
+								if (identifier == "NETSCAPE")
+									s.Position += 5;
+								else
+									readSubblock(s);
+								break;
+							default :
+								readSubblock(s);
+								break;
+						}
+						break;
+					case BlockTypes.IMAGE_DESCRIPTOR:
+						frames.Add(new GifFrame(s, size));
+						break;
+					case BlockTypes.TRAILER :
+						return frames.ToArray();
+					default :
+						Console.WriteLine("Bad data");
+						break;
+				}
+
+				// Shouldnt happen but if Trailer is missing at EOF we should bail.
+				if (s.Position >= s.Length)
+					break;
+			}
+
+			return frames.ToArray();
+		}
+
+		public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
+		{
+			if (!IsGif(s))
+			{
+				frames = null;
+				return false;
+			}
+
+			frames = ParseFrames(s);
+			return true;
+		}
+
+		#region Helper functions
+
+		static int bitsToNum(byte[] array)
+		{
+			var value = (int)(array[0]);
+			for (var i = 1; i < array.Length; i++)
+			{
+				value = value * 2 + array[i];
+			}
+
+			return value;
+		}
+
+		// Each entry is 3 bytes in RGB
+		static byte[] parseColorTable(Stream s, int entries)
+		{
+			var ct = new List<byte>();
+			for (var i = 0; i < entries; i++)
+			{
+				ct.AddRange(s.ReadBytes(3));
+			}
+
+			return ct.ToArray();
+		}
+
+		static byte[] readSubblock(Stream s)
+		{
+			var data = new List<byte>();
+
+			while (true)
+			{
+				var ssize = s.ReadUInt8();
+				if (ssize == 0)
+					break;
+
+				data.AddRange(s.ReadBytes(ssize));
+			}
+
+			return data.ToArray();
+		}
+
+		static byte[] decode(int minCodeSize, byte[] data)
+		{
+			var pos = 0;
+			var code = 0;
+			var last = 0;
+
+			var decrypt = new Action<int>((size) =>
+			{
+				code = 0;
+				for (var i = 0; i < size; i++)
+				{
+					if ((data[pos >> 3] & (1 << (pos & 7))) != 0)
+						code |= 1 << i;
+					pos++;
+				}
+			});
+
+			var output = new List<byte>();
+
+			var clearCode = 1 << minCodeSize;
+			var eoiCode = clearCode + 1;
+
+			var codeSize = minCodeSize + 1;
+
+			var dict = new List<byte[]>();
+
+			while (true)
+			{
+				last = code;
+				decrypt(codeSize);
+
+				if (code == clearCode)
+				{
+					dict.Clear();
+					codeSize = minCodeSize + 1;
+					for (var i = 0; i < clearCode; i++)
+						dict.Add(new byte[]{(byte)i});
+					dict.Add(new byte[]{});
+					dict.Add(null);
+					continue;
+				}
+				if (code == eoiCode)
+					break;
+				if (code < dict.Count)
+				{
+					if (last != clearCode)
+						dict.Add(dict.ElementAt(last).Concat(new byte[]{dict.ElementAt(code)[0]}).ToArray());
+				}
+				else
+				{
+					if (code != dict.Count)
+						throw new Exception("Invalid LZW code");
+					var element = dict.ElementAt(last);
+					dict.Add(element.Concat(new byte[]{element[0]}).ToArray());
+				}
+
+				output = output.Concat(dict.ElementAt(code)).ToList();
+
+				if (dict.Count == (1 << codeSize) && codeSize < 12)
```

Comment:
I feel like it would make this less readable, the shift has hight precedence in C# but I feel like it makes more sense to have it.